swap-intent/src/ft.rs

```diff
@@ -0,0 +1,65 @@
+use defuse_contracts::intents::swap::{Asset, FtAmount, SwapIntentAction, SwapIntentError};
+use near_contract_standards::fungible_token::{core::ext_ft_core, receiver::FungibleTokenReceiver};
+use near_sdk::{
+    env, json_types::U128, near, serde_json, AccountId, NearToken, Promise, PromiseOrValue,
+};
+
+use crate::{SwapIntentContractImpl, SwapIntentContractImplExt};
+
+#[near]
+impl FungibleTokenReceiver for SwapIntentContractImpl {
+    fn ft_on_transfer(
+        &mut self,
+        sender_id: AccountId,
+        amount: U128,
+        msg: String,
+    ) -> PromiseOrValue<U128> {
+        self.internal_ft_on_transfer(sender_id, amount, msg)
+            .unwrap()
+    }
+}
+
+impl SwapIntentContractImpl {
+    fn internal_ft_on_transfer(
+        &mut self,
+        sender_id: AccountId,
+        amount: U128,
+        msg: impl AsRef<str>,
+    ) -> Result<PromiseOrValue<U128>, SwapIntentError> {
+        if amount.0 == 0 {
+            return Err(SwapIntentError::ZeroAmount);
+        }
+        let action = serde_json::from_str(msg.as_ref()).map_err(SwapIntentError::JSON)?;
+
+        let received = Asset::Ft(FtAmount {
+            token: env::predecessor_account_id(),
+            amount: amount.0,
+        });
+
+        Ok(match action {
+            SwapIntentAction::Create(create) => {
+                self.create_intent(sender_id, received, create)?;
+                // intent was successfully created, do not refund
+                PromiseOrValue::Value(0.into())
```

Comment:
Because `create_intent` is used to create intents where "asset_in"  can be one of: native/ft/nft. All of these interfaces require different return values (e.g. nft doesn't have a concept of `amount`), so we handle them differently from a higher level caller. See [native](https://github.com/defuse-is-near/defuse-contracts/blob/0c787bb84a958b1b248b3110f34a19da33512215/swap-intent/src/native.rs#L28), [ft](https://github.com/defuse-is-near/defuse-contracts/blob/0c787bb84a958b1b248b3110f34a19da33512215/swap-intent/src/ft.rs#L43), [nft](https://github.com/defuse-is-near/defuse-contracts/blob/0c787bb84a958b1b248b3110f34a19da33512215/swap-intent/src/nft.rs#L43)